virttest/utils_conn.py

```diff
@@ -1946,3 +1979,181 @@ def conn_setup(self):
+            'remove_my_rpcworker_policy', False)
+        init_dict['remove_qemu_kvm_policy'] = init_dict.get(
+            'remove_qemu_kvm_policy', False)
+        super(UNIXSocketConnection, self).__init__(init_dict)
+
+        self.pmsocat_path = os.path.join(data_dir.get_shared_dir(),
+                                         'scripts/pmsocat', 'pmsocat36.py')
+        self.remote_pmsocat_path = os.path.join('/tmp', 'pmsocat36.py')
+        remote_old.scp_to_remote(self.server_ip, '22', self.server_user,
+                                 self.server_pwd, self.pmsocat_path,
+                                 self.remote_pmsocat_path)
+
+    def clear_pmsocat(self, ignore_status=False):
+        """
+        Clear pmsocat processes.
+
+        :param ignore_status: Whether to raise an exception when command fails
+        """
+        for session in [self.client_session, self.server_session]:
+            cmd = "pkill -f pmsocat36"
+            status, output = session.cmd_status_output(cmd)
+            if not ignore_status:
+                if status:
+                    raise ConnCmdClientError(cmd, output)
+
+    def conn_recover(self):
+        """
+        Clean up for Unix socket connection.
+
+        (1). Remove firewall ports
+        (2). Stop pmsocat process
+        (3). Remove my-rpcworker policy
+        """
+        if self.setup_on_local:
+            tcp2unix_session = self.server_session
+            dest_pmsocat_path = self.remote_pmsocat_path
+            unix2tcp_session = self.client_session
+        else:
+            tcp2unix_session = self.client_session
+            dest_pmsocat_path = self.pmsocat_path
+            unix2tcp_session = self.server_session
+
+        firewall_cmd = utils_iptables.Firewall_cmd(tcp2unix_session)
+        for port_to_del in [self.desturi_port, self.migrateuri_port, self.disks_uri_port]:
+            firewall_cmd.remove_port(port_to_del, 'tcp')
+        self.clear_pmsocat(ignore_status=True)
+        tcp2unix_session.cmd("rm -rf %s" % dest_pmsocat_path)
+        if self.remove_my_rpcworker_policy:
+            logging.debug("Remove my-rpcworker policy.")
+            cmd = "semodule -X 300 -r my-rpcworker"
+            status, output = unix2tcp_session.cmd_status_output(cmd)
+            if status:
+                raise ConnCmdClientError(cmd, output)
+        if self.remove_qemu_kvm_policy:
+            logging.debug("Remove qemu-kvm policy.")
+            cmd = "semodule -r qemu-kvm"
+            status, output = unix2tcp_session.cmd_status_output(cmd)
+            if status:
+                raise ConnCmdClientError(cmd, output)
+        logging.debug("UNIX sockets recover successfully.")
+
+    def conn_setup(self):
+        """
+        Setup unix socket connection.
+
+        (1). Start a unix proxy that listens on tcp socket
+        (2). Active rpcworker policy
+        (3). Start a proxy that listens on tcp socket
+        (4). Add firewall ports
+        """
```

Comment:
@Yingshun This is a rather large method. I'm wondering if you considered putting parts (1-4) into smaller testable functions and call them here?

Reply:
Yes, done.